pkg/maintainer/spv/redemptions.go

```diff
@@ -0,0 +1,312 @@
+		if err != nil {
+			return nil, fmt.Errorf("failed to get wallet: [%v]", err)
+		}
+
+		if wallet.State != tbtc.StateLive &&
+			wallet.State != tbtc.StateMovingFunds {
+			// The wallet can only submit redemption proofs if it's `Live` or
+			// `MovingFunds`. If the state is different skip it.
+			logger.Infof(
+				"skipped proving redemption transactions for wallet [%x] "+
+					"because of wallet state [%v]",
+				walletPublicKeyHash,
+				wallet.State,
+			)
+			continue
+		}
+
+		walletTransactions, err := btcChain.GetTransactionsForPublicKeyHash(
+			walletPublicKeyHash,
+			transactionLimit,
+		)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to get transactions for wallet: [%v]",
+				err,
+			)
+		}
+
+		for _, transaction := range walletTransactions {
+			isUnproven, err :=
+				isUnprovenRedemptionTransaction(
+					transaction,
+					walletPublicKeyHash,
+					btcChain,
+					spvChain,
+				)
+			if err != nil {
+				return nil, fmt.Errorf(
+					"failed to check if transaction is an unproven redemption "+
+						"transaction: [%v]",
+					err,
+				)
+			}
+
+			if isUnproven {
+				unprovenRedemptionTransactions = append(
+					unprovenRedemptionTransactions,
+					transaction,
+				)
+			}
+		}
+	}
+
+	return unprovenRedemptionTransactions, nil
+}
+
+func isUnprovenRedemptionTransaction(
+	transaction *bitcoin.Transaction,
+	walletPublicKeyHash [20]byte,
+	btcChain bitcoin.Chain,
+	spvChain Chain,
+) (bool, error) {
+	// If the transaction does not have exactly one input, it cannot be a
+	// redemption transaction.
+	if len(transaction.Inputs) != 1 {
+		return false, nil
+	}
+
+	singleInput := transaction.Inputs[0]
+
+	// Check whether the single input is the current wallet main UTXO.
+	isMainUtxo, err := isInputCurrentWalletsMainUTXO(
+		singleInput.Outpoint.TransactionHash,
+		singleInput.Outpoint.OutputIndex,
+		walletPublicKeyHash,
+		btcChain,
+		spvChain,
+	)
+	if err != nil {
+		return false, fmt.Errorf(
+			"failed to check if input is the main UTXO",
+		)
+	}
+
+	// If the single input is not the current main UTXO of the wallet, the
+	// transaction is either a redemption transaction that is already
+	// proven or it's not a redemption transaction at all.
+	if !isMainUtxo {
+		return false, nil
+	}
+
+	changeFound := false
+
+	// Look at the transaction's outputs. All the outputs must be pending
+	// redemption requests, except for one optional change output.
+	for _, output := range transaction.Outputs {
+		// First, check if the given output is a change (if it wasn't
+		// found yet).
+		if !changeFound {
+			isChange, err := isWalletChangeOutput(walletPublicKeyHash, output)
+			if err != nil {
+				return false, fmt.Errorf(
+					"failed to check if output is wallet change: [%v]",
+					err,
+				)
+			}
+
+			if isChange {
+				changeFound = true
+				continue
+			}
+		}
+
+		// If the given output is not a change, it must be a pending redemption
+		// request.
```

Comment:
According to [some checks in Bridge](https://github.com/keep-network/tbtc-v2/blob/716bf066c2cae0ccf8fe16b71ac43dac2bb7b708/solidity/contracts/bridge/Redemption.sol#L925) the output of a redemption transaction can also be a timed-out redemption.
If the wallet made a redemption transaction handling such timed-out redemption requests we should probably handle it.

Reply:
I think we should not submit such proofs automatically. This is an abnormal situation that should not happen and this condition lives in the Bridge just in case. It will be safer if the maintainer fails and a human can assess the situation and submit the proof manually if needed.